### PR TITLE
Implement SSD Model Architecture for WSL

### PR DESCRIPTION
**Pull request by farazali7**
This PR is for implementing the general model architecture used for our weakly supervised B-line detection project, as described in the [paper by Zhao et. al (2021)](https://www.sciencedirect.com/science/article/abs/pii/S0925231221004215). The functionality implemented thus far in this PR is not at the point of training yet - it is mainly centered around the model architecture, loss function, and object detection utilities. A subsequent PR for the pseudo ground-truth module and training orchestration will be made. 

**TLDR for model architecture from the paper:**

image

- The model operates on embeddings extracted from a pretrained backbone; in our case it is the trained AB-line classifier
- The embeddings go through a Pseudo Ground-Truth Module (PGM) to extract a saliency map which is used to guide the training of the detection network; in our case it is currently heatmaps obtained from Grad-CAM++
- The detection occurs in the Single-shot Detection Module (SDM); this is comprised of a Convolution Prediction Module (CPM and a Deconvolution Prediction Module (DPM)
- The CPM is a stack of three convolutional blocks, the first two of which are shared with the last two conv blocks in the pretrained backbone. The job of the CPM is to filter out anchor boxes that contain no objects, while regressing on the positive sample anchor box positions to improve their location.
- The DPM, made up of three deconvolutional layers, receives the filtered boxes from the CPM and performs standard classification and localization on them

Summary of changes:
- Restructure repository into classification and detection folders
- Implement base network architecture including connection of pretrained model to SSD as defined in paper
- Implement loss functionality as defined in paper
- Add bounding box and SSD utility helper functions
- Add function for extracting bounding box coordinates from heatmaps

For `extract_bbox_from_heatmap`:
I would like to get opinions on the potential ways to filter some of the bboxes. Currently, there is just a filter based on number of contour points (arbitrarily set to 20 for now), but I am not in love with that idea - so others are much appreciated! I have also added some parameters to toggle thresholds and methods use to potentially improve the quality of the obtained bounding boxes before contouring, such as Otsu's Thresholding and dilation. However, these sometimes result in generating additional incorrect boxes - so there is tradeoff. See example below of A-line GT class:

Binarization with threshold = 127 (misses A-line seen around 50x, 40y on grayscale heatmap):
image

Binarization with Otsu's thresholding:
image

Binarization with Otsu's thresholding + dilation (kernel is shape 1x5 so it spreads mask out horizontally):
image

### CODE REVIEW

File: src/detection/losses/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from .smooth_l1_loss import SMOOTH_L1_LOSS
```
Comment (hmathlee) [inline src/detection/losses/__init__.py:1]: What are your thoughts on putting these loss classes into a single `losses.py` file?

File: config.yml [hunk 2]
```diff
@@ -115,7 +116,43 @@ HPARAMS:
...
+    LR_EXTRACT: 0.0003
+    LR_FINETUNE: 0.0000093
+    DROPOUT: 0.45
+    EPOCHS: 5
+    FROZEN_LAYERS: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
+    EXTRA_BOX_FOR_AR1: False
+    MIN_SCALE: 0.2
+    MAX_SCALE: 0.9
+    VARIANCES: [0.1, 0.1, 0.2, 0.2]
+    PREDICTION_LAYERS:
```
Comment (hmathlee) [inline config.yml:129]: Could add some comments here (??)

File: README.md [hunk 3]
```diff
@@ -37,12 +37,12 @@ _A deep learning solution for the classification of normal versus abnormal lung
...
+5. Execute [_train.py_](src/classification/train.py) to train your chosen model on your
    preprocessed data. The trained model will be serialized within
    _results/models/_, and its filename will resemble the following
    structure: _model{yyyymmdd-hhmmss}.h5_, where _{yyyymmdd-hhmmss}_ is the current
    time.  
-   Note: When [_train.py_](src/train.py) is executed, the project root should be the working directory.
+   Note: When [_train.py_](src/classification/train.py) is executed, the project root should be the working directory.
 6. Navigate to _results/logs/_ to see the tensorboard log files. The folder name will
    be _{yyyymmdd-hhmmss}_.  These logs can be used to create a [tensorboard](https://www.tensorflow.org/tensorboard)
    visualization of the training results.
```
Comment (hmathlee) [inline README.md:48]: Nice explanations!

**Review by hmathlee — COMMENTED** [review]
Amazing work Faraz! I left some very minor comments.

File: src/detection/losses/softmax_ce_loss.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import tensorflow as tf
+from tensorflow.keras.losses import CategoricalCrossentropy, BinaryCrossentropy
+
+
+class SOFTMAX_CE_LOSS:
+    def compute(self, y_true, y_pred, ce_type='binary'):
```
[the remark this answers is not in the document]
Comment (farazali7) [reply src/detection/losses/softmax_ce_loss.py]: Yes, I've added this and `smoothing_point` you noted below to the config file and reconfigured the `SSD_LOSS` class to initialize with both instead.

File: src/detection/utils/bbox_utils/perform_nms.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import tensorflow as tf
+
+
+def perform_nms(single_class, max_output_size, iou_threshold):
```
[the remark this answers is not in the document]
Comment (farazali7) [reply src/detection/utils/bbox_utils/perform_nms.py:4]: Yes! Not sure how I missed this... good catch!

File: src/detection/losses/__init__.py (hunk 1 above)
Comment (farazali7) [reply src/detection/losses/__init__.py]: That's probably more concise - I've consolidated them into a single file in latest update. (Also is how Keras does them!)

File: src/detection/losses/ssd_loss.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        :param y_true: Tensor of pseudo ground-truth labels from both CPM and DPM
+        :param y_pred: Tensor of predicted labels from both CPM and DPM
+        :return: Sum of total L1 regression and classification loss for network
+        '''
+        # Labels and preds have shape: (batch size, module, boxes, n+1 classes + 4 loc preds + 8 anchor values)
+        cpm_bbox_true = y_true[:, 0, :, :]
+        cpm_bbox_pred = y_pred[:, 0, :, :]
+
+        dpm_bbox_true = y_true[:, 1, :, :]
+        dpm_bbox_pred = y_pred[:, 1, :, :]
```
[the remark this answers is not in the document]
Comment (bvanberl) [reply src/detection/losses/ssd_loss.py]: Yeah I wouldn't use newaxis() here - I think it's meant for adding >= 1 new dim to an array.

**Review by bvanberl — APPROVED** [review]
Stellar work @farazali7 . Very well designed and documented as well.

**Review by bvanber — APPROVED** [review]

File: src/detection/utils/ssd_utils/extract_bbox_from_heatmap.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    # Set threshold to half of max val if none provided
+    if threshold is None:
+        threshold = max_val // 2
+
+    gray_heatmap = heatmap[:, :, 0] if is_grayscale else cv2.cvtColor(heatmap, cv2.COLOR_RGB2GRAY)
+
+    ret, thresh = cv2.threshold(gray_heatmap,
+                                thresh=threshold,
+                                maxval=max_val,
+                                type=thresh_type + (int(perform_otsu)*cv2.THRESH_OTSU))
```
[the remark this answers is not in the document]
Comment (farazali7) [reply src/detection/utils/ssd_utils/extract_bbox_from_heatmap.py:31]: `thresh_type` is an enum from `cv::ThresholdTypes` from 0 to 7. The implementation ([linked here](https://github.com/opencv/opencv/blob/139c44377032f58849bbab2ae454fcf14e89d762/modules/imgproc/src/thresh.cpp#L1540-L1558)) uncovers the presence of the additional Otsu's thresholding flag through bitwise subtraction - so any of the `thresh_type`s can be used with the Otsu's flag. The `int(perform_otsu)` is just a switch I added for turning the flag on/off.
